File: src/test/java/com/vjiki/music/repository/ChatRepositoryIntegrationTest.java
package com.vjiki.music.repository;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import java.util.Optional;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;

import com.vjiki.music.MusicApplication;
import com.vjiki.music.config.TestContainersConfig;
import com.vjiki.music.entity.AccessLevel;
import com.vjiki.music.entity.AuthProvider;
import com.vjiki.music.entity.Chat;
import com.vjiki.music.entity.ChatParticipant;
import com.vjiki.music.entity.ChatType;
import com.vjiki.music.entity.ParticipantRole;
import com.vjiki.music.entity.User;

@DataJpaTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@ContextConfiguration(classes = {MusicApplication.class, TestContainersConfig.class})
class ChatRepositoryIntegrationTest {

    @DynamicPropertySource
    static void registerDataSourceProperties(DynamicPropertyRegistry registry) {
        registry.add("spring.datasource.url", TestContainersConfig.POSTGRES_CONTAINER::getJdbcUrl);
        registry.add("spring.datasource.username", TestContainersConfig.POSTGRES_CONTAINER::getUsername);
        registry.add("spring.datasource.password", TestContainersConfig.POSTGRES_CONTAINER::getPassword);
        registry.add("spring.datasource.hikari.data-source-properties.ssl", () -> "false");
        registry.add("spring.datasource.hikari.data-source-properties.sslmode", () -> "disable");
        // "update" rather than "create-drop": this slice shares one container with the
        // @SpringBootTest contexts, and dropping tables would break whichever context is alive.
        registry.add("spring.jpa.hibernate.ddl-auto", () -> "update");
    }

    @Autowired
    private ChatRepository chatRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private ChatParticipantRepository chatParticipantRepository;

    private User saveUser(String email, String nickname) {
        return userRepository.save(User.builder()
                .email(email)
                .nickname(nickname)
                .provider(AuthProvider.LOCAL)
                .accessLevel(AccessLevel.USER)
                .isActive(true)
                .createdBy("system")
                .modifiedBy("system")
                .build());
    }

    @Nested
    @DisplayName("ChatRepository")
    class ChatRepositoryTests {

        @Test
        @DisplayName("should save and find chat by id")
        void shouldSaveAndFindChatById() {
            Chat chat = Chat.builder()
                    .type(ChatType.DIRECT)
                    .build();

            Chat saved = chatRepository.save(chat);
            Optional<Chat> found = chatRepository.findById(saved.getId());

            assertThat(found).isPresent();
            assertThat(found.get().getType()).isEqualTo(ChatType.DIRECT);
        }

        @Test
        @DisplayName("should find chats by user id")
        void shouldFindChatsByUserId() {
            User user1 = saveUser("user1@example.com", "user1");
            saveUser("user2@example.com", "user2");

            Chat chat1 = chatRepository.save(Chat.builder()
                    .type(ChatType.DIRECT)
                    .build());
            Chat chat2 = chatRepository.save(Chat.builder()
                    .type(ChatType.GROUP)
                    .build());

            chatParticipantRepository.save(ChatParticipant.builder()
                    .chatId(chat1.getId())
                    .userId(user1.getId())
                    .chat(chat1)
                    .user(user1)
                    .role(ParticipantRole.MEMBER)
                    .isMuted(false)
                    .build());
            chatParticipantRepository.save(ChatParticipant.builder()
                    .chatId(chat2.getId())
                    .userId(user1.getId())
                    .chat(chat2)
                    .user(user1)
                    .role(ParticipantRole.MEMBER)
                    .isMuted(false)
                    .build());

            List<Chat> chats = chatRepository.findChatsByUserId(user1.getId());

            assertThat(chats).hasSize(2);
        }
    }
}

File: bin/test/com/vjiki/music/repository/ChatRepositoryIntegrationTest.kt
package com.vjiki.music.repository

import com.vjiki.music.config.TestContainersConfig
import com.vjiki.music.entity.*
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.extensions.spring.SpringExtension
import io.kotest.matchers.shouldBe
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.test.context.ContextConfiguration
import java.util.*

@DataJpaTest
@ContextConfiguration(classes = [TestContainersConfig::class])
open class ChatRepositoryIntegrationTest : DescribeSpec() {

    override fun extensions() = listOf(SpringExtension)

    @Autowired
    lateinit var chatRepository: ChatRepository

    @Autowired
    lateinit var userRepository: UserRepository

    @Autowired
    lateinit var chatParticipantRepository: ChatParticipantRepository

    init {
        describe("ChatRepository") {
        it("should save and find chat by id") {
            val chat = Chat(
                type = ChatType.DIRECT
            )

            val saved = chatRepository.save(chat)
            val found = chatRepository.findById(saved.id)

            found.isPresent shouldBe true
            found.get().type shouldBe ChatType.DIRECT
        }

        it("should find chats by user id") {
            val user1 = userRepository.save(User(
                email = "user1@example.com",
                nickname = "user1",
                provider = AuthProvider.LOCAL,
                accessLevel = AccessLevel.USER,
                isActive = true,
                createdBy = "system",
                modifiedBy = "system"
            ))
            val user2 = userRepository.save(User(
                email = "user2@example.com",
                nickname = "user2",
                provider = AuthProvider.LOCAL,
                accessLevel = AccessLevel.USER,
                isActive = true,
                createdBy = "system",
                modifiedBy = "system"
            ))

            val chat1 = chatRepository.save(Chat(
                type = ChatType.DIRECT
            ))
            val chat2 = chatRepository.save(Chat(
                type = ChatType.GROUP
            ))

            chatParticipantRepository.save(ChatParticipant(
                chatId = chat1.id,
                userId = user1.id,
                chat = chat1,
                user = user1,
                role = ParticipantRole.MEMBER,
                isMuted = false
            ))
            chatParticipantRepository.save(ChatParticipant(
                chatId = chat2.id,
                userId = user1.id,
                chat = chat2,
                user = user1,
                role = ParticipantRole.MEMBER,
                isMuted = false
            ))

            val chats = chatRepository.findChatsByUserId(user1.id)

            chats.size shouldBe 2
        }
    }
    }
}

